### PR TITLE
HPCC-31153 Document LDAP Initialization

### PR DESCRIPTION
**Pull request by kenrowland**
Added writeup for LDAP initialization

Signed-Off-By: Kenneth Rowland kenneth.rowland@lexisnexisrisk.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: devdoc/LDAPSecurityManager.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+#### AD Hosts
+The LDAP Security Manager supports using multiple ADs. The FDQN or IP address of each AD host is read from configuration
+data and stored internally. The source is a comma separated list stored in the _ldapAddress_ config value. Each entry
+is added to a pool of ADs
+
+**Note that all ADs are expected to use the same credentials and have the same configuration**
+
+#### AD Credentials
+AD credentials consist of a username and password. The LDAP security manager uses these to perform all operations
+on behalf of uses and components in the cluster. There are three potential sources for credentials.
```
Comment (JamesDeFabia) [inline devdoc/LDAPSecurityManager.md:26]: uses? did you mean users?

File: devdoc/LDAPSecurityManager.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+1. A Kubernetes secret. If the _ldapAdminSecretKey_ config value is set, but _ldapAdminVaultId_ is not (see 2) then
+the AD credentials are retrieve as Kubernetes secrets.
```
Comment (JamesDeFabia) [inline devdoc/LDAPSecurityManager.md:29]: sp retrieve s/b retrieved

File: devdoc/LDAPSecurityManager.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+2. If both _ldapAdminSecretKey_ and _ldapAdminVaultId_ config values are present, the AD credentials are retrieved
+from the vault.
+3. Hardcoded values from the _SystemCommonName_ and _systemPassword_ config values stored in the environment.xml file.
+
+As stated above, when multiple ADs are in use, the configuration of each must be the same. This includes credentials.
+
+### Retrieve Server Information from the AD ###
+During initialization, the security manager begins incrementing through the set of defines ADs until it is able to
```
Comment (JamesDeFabia) [inline devdoc/LDAPSecurityManager.md:37]: defines s/b defined

File: devdoc/LDAPSecurityManager.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+The following process is used when an LDAP connection is needed.
+
+First, the connection pool is searched for a free connection. If found and valid, the connection is returned. A 
+connection is considered free if no one is using it and valid if the AD can be accessed. If no valid free connections 
+are found, a new uninitialized connection is created.
+
+For a new connection, an attempt is made to connect to each AD starting with the current. See below for how ADs 
+are cycled when a connection fails. For each AD, as it cycles through, connection attempts are retried with a short
+delay between each. if unable to connect, the AD host is marked rejected and the next is attempted. Once a new
```
Comment (JamesDeFabia) [inline devdoc/LDAPSecurityManager.md:65]: if s/b If (begins  sentence)

**Review by JamesDeFabia — CHANGES_REQUESTED** [review]
A few minor comments inline

File: devdoc/LDAPSecurityManager.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
+# LDAP Security Manager Init
+This document covers the main steps taken by the LDAP Security manager when initializing an AD via LDAP. An important
+distinction is that AD, Active Directory, is the store where permissions, users, password, resource, etc. are stored.
```
Comment (afishbeckln) [inline devdoc/LDAPSecurityManager.md:3]: passwords, resources

File: devdoc/LDAPSecurityManager.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+LDAP is the protocol used to access the AD. The term LDAP is generally overused to refer to both.
+
+## LDAP Instances ##
+It is important to note that each service and or component using the LDAP security manager gets its
```
Comment (afishbeckln) [inline devdoc/LDAPSecurityManager.md:7]: and/or

File: devdoc/LDAPSecurityManager.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+#### AD Credentials
+AD credentials consist of a username and password. The LDAP security manager uses these to perform all operations
+on behalf of uses and components in the cluster. There are three potential sources for credentials.
+
+1. A Kubernetes secret. If the _ldapAdminSecretKey_ config value is set, but _ldapAdminVaultId_ is not (see 2) then
+the AD credentials are retrieve as Kubernetes secrets.
+2. If both _ldapAdminSecretKey_ and _ldapAdminVaultId_ config values are present, the AD credentials are retrieved
+from the vault.
+3. Hardcoded values from the _SystemCommonName_ and _systemPassword_ config values stored in the environment.xml file.
```
Comment (afishbeckln) [inline devdoc/LDAPSecurityManager.md:32]: Is the case for each "S" correct?  It's just inconsistent so I thought I'd ask.

File: devdoc/LDAPSecurityManager.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
+# LDAP Security Manager Init
+This document covers the main steps taken by the LDAP Security manager when initializing an AD via LDAP. An important
```
Comment (afishbeckln) [inline devdoc/LDAPSecurityManager.md:2]: "when initializing an AD which will be accessed via LDAP"?

**Review by afishbeckln — CHANGES_REQUESTED** [review]
A few minor comments, but no major changes.

File: devdoc/LDAPSecurityManager.md (hunk 7 above)
Comment (kenrowland) [reply devdoc/LDAPSecurityManager.md:32]: It is lower case 's' for systemCommonName.

File: devdoc/LDAPSecurityManager.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+## LDAP Instances ##
+Each service and/or component using the LDAP security manager gets its own instance of the security manager. This 
+includes a unique connection pool (see below). All operations described apply to each LDAP instance.
+
+## Initialization Steps
+The following sections cover the main steps taken during initialization
+
+### Load Configuration
+The following items are loaded from the configuration
```
Comment (JamesDeFabia) [inline devdoc/LDAPSecurityManager.md:14]: I think this should end in a colon (:)

File: devdoc/LDAPSecurityManager.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+#### AD Credentials
+AD credentials consist of a username and password. The LDAP security manager users these to perform all operations
+on behalf of users and components in the cluster. There are three potential sources for credentials.
+
+1. A Kubernetes secret. If the _ldapAdminSecretKey_ config value is set, but _ldapAdminVaultId_ is not (see 2) then
+the AD credentials are retrieved as Kubernetes secrets.
+2. If both _ldapAdminSecretKey_ and _ldapAdminVaultId_ config values are present, the AD credentials are retrieved
+from the vault.
+3. Hardcoded values from the _systemCommonName_ and _systemPassword_ config values stored in the environment.xml file.
+
```
Comment (JamesDeFabia) [inline devdoc/LDAPSecurityManager.md:32]: Should we warn that #3 is not a secure method of storage?

File: devdoc/LDAPSecurityManager.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+As stated above, when multiple ADs are in use, the configuration of each must be the same. This includes credentials.
+
+### Retrieve Server Information from the AD ###
+During initialization, the security manager begins incrementing through the set of defined ADs until it is able to
```
Comment (JamesDeFabia) [inline devdoc/LDAPSecurityManager.md:36]: incrementing? Did you mean traversing or iterating?

File: devdoc/LDAPSecurityManager.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+During initialization, the security manager begins incrementing through the set of defined ADs until it is able to 
+connect and retrieve information from an AD. Once retrieved, the information is used for all ADs (see statement 
+above about all ADs being the same). The accessed AD is marked as the current AD and no other ADs are accessed
+during initialization.
+
+The retrieved information is used to verify the AD type so the security manager adjusts for variations between 
+types. Additionally, defined DNs may be adjusted to match AD type requirements.
+
+# LDAP Operations
+This section covers what it does....
```
Comment (JamesDeFabia) [inline devdoc/LDAPSecurityManager.md:45]: An ellipse is only three dots. But maybe a period is better here.

File: devdoc/LDAPSecurityManager.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+until the max allowed is reached. 
+
+The following process is used when an LDAP connection is needed.
+
+First, the connection pool is searched for a free connection. If found and valid, the connection is returned. A 
+connection is considered free if no one is using it and valid if the AD can be accessed. If no valid free connections 
+are found, a new uninitialized connection is created.
+
+For a new connection, an attempt is made to connect to each AD starting with the current. See **Handling AD Hosts**
+below for how ADs are cycled when a connection fails.For each AD, as it cycles through, connection attempts are retried
```
Comment (JamesDeFabia) [inline devdoc/LDAPSecurityManager.md:63]: need a space between period and For

File: devdoc/LDAPSecurityManager.md [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+### Handling AD Hosts
+The manager keeps a list of AD hosts and the index of the current host. The current host is used for all AD
+operations until there is a failure. At that time the manager marks the host as "rejected" and moves to the 
+next host using a round-robin scheme. 
+
+## User Authentication & Authorization
+This is the process of verifying the identity of a user. It also verifies authorization for an optional resource list.
+The security manager _authorize_ interface method performs this task. It accepts a user and optional resource list.
+Currently only a username and password are required to authenticating the user. Once authenticated, the user is
```
Comment (JamesDeFabia) [inline devdoc/LDAPSecurityManager.md:80]: authenticating s/b authenticate

File: devdoc/LDAPSecurityManager.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+This is the process of verifying the identity of a user. It also verifies authorization for an optional resource list.
+The security manager _authorize_ interface method performs this task. It accepts a user and optional resource list.
+Currently only a username and password are required to authenticating the user. Once authenticated, the user is 
+checked to make sure they are authorized to access the optional resources listed. As an example, users logging into
+ECLWatch require **SmcAccess** permission. It is possible a user authenticates, but is not authorized for one or more
+optional resources. In this case, the _authorize_ call will fail.
+
+The LDAP security manager may cache user credentials in order to speed up authentication (not resource authorization)
+when a user attempts multiple authorization calls. The cache is 
+
```
Comment (JamesDeFabia) [inline devdoc/LDAPSecurityManager.md:87]: "The cache is" _____________.
Missing the rest of the sentence

File: devdoc/LDAPSecurityManager.md [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+### Per Request Authentication ###
+While not part of LDAP directly, every request from ECLWatch is authenticated to make sure the calling user is valid.
```
Comment (JamesDeFabia) [inline devdoc/LDAPSecurityManager.md:90]: ECL Watch

**Review by JamesDeFabia — CHANGES_REQUESTED** [review]
Few comments inline

File: devdoc/LDAPSecurityManager.md (hunk 10 above)
Comment (kenrowland) [reply devdoc/LDAPSecurityManager.md:32]: It's not really the purpose of the document. I think that is something reserved for end user docs discussing how to setup a secure system.

File: devdoc/LDAPSecurityManager.md (hunk 11 above)
Comment (kenrowland) [reply devdoc/LDAPSecurityManager.md:36]: Incrementing is fine. It's an array internally.

File: devdoc/LDAPSecurityManager.md (hunk 12 above)
Comment (kenrowland) [reply devdoc/LDAPSecurityManager.md:45]: That section should have been removed. I removed it.

File: devdoc/LDAPSecurityManager.md (hunk 15 above)
Comment (kenrowland) [reply devdoc/LDAPSecurityManager.md:87]: This last section should have been removed as well. It is now covered in another document.

File: devdoc/LDAPSecurityManager.md (hunk 10 above)
Comment (JamesDeFabia) [reply devdoc/LDAPSecurityManager.md:32]: fair enough

**Review by JamesDeFabia — APPROVED** [review]
Looks good from my POV

**Review by afishbeckln — APPROVED** [review]

**Comment by kenrowland** [comment]
@ghalliday Please merge

**Pull request merged**